src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -87,115 +109,119 @@ EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, monodroid_sh
 		if (compression_method != 0)
 			continue;
 
-		if (strncmp (prefix, file_name, prefix_len) != 0)
+		if (strncmp (prefix, entry_name.get (), prefix_len) != 0)
 			continue;
 
 #if defined (NET6)
 		if (application_config.have_runtime_config_blob && !runtime_config_blob_found) {
-			if (utils.ends_with (file_name, SharedConstants::RUNTIME_CONFIG_BLOB_NAME)) {
+			if (utils.ends_with (entry_name, SharedConstants::RUNTIME_CONFIG_BLOB_NAME)) {
 				runtime_config_blob_found = true;
-				runtime_config_blob_mmap = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+				runtime_config_blob_mmap = md_mmap_apk_file (fd, data_offset, file_size, entry_name.get (), apk_name);
 				continue;
 			}
 		}
 #endif // def NET6
 
 		// assemblies must be 4-byte aligned, or Bad Things happen
 		if ((data_offset & 0x3) != 0) {
-			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", file_name, data_offset);
+			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", entry_name.get (), data_offset);
 			log_fatal (LOG_ASSEMBLY, "You MUST run `zipalign` on %s\n", strrchr (apk_name, '/') + 1);
 			exit (FATAL_EXIT_MISSING_ZIPALIGN);
 		}
 
-		bool entry_is_overridden = !should_register (strrchr (file_name, '/') + 1);
+		const char *last_slash = utils.find_last (entry_name, '/');
+		bool entry_is_overridden = last_slash == nullptr ? false : !should_register (last_slash + 1);
 
-		if ((utils.ends_with (file_name, ".pdb") || utils.ends_with (file_name, ".mdb")) &&
+		if ((utils.ends_with (entry_name, ".pdb") || utils.ends_with (entry_name, ".mdb")) &&
 				register_debug_symbols &&
 				!entry_is_overridden &&
-				bundled_assemblies != nullptr) {
-			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, file_name, apk_name);
-			if (register_debug_symbols_for_assembly (file_name, (bundled_assemblies) [bundled_assemblies_count - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
+				bundled_assembly_index >= 1) {
+			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, entry_name.get (), apk_name);
+			if (register_debug_symbols_for_assembly (entry_name, (bundled_assemblies) [bundled_assembly_index - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
```

Comment:
Why parens around `(bundled_assemblies)`?  (Yes, it *previously* had them, but it shouldn't need them *now*…)

I also wonder if this is a potential future "bug," *requiring* that the `.dll` come *before* the `.pdb`.  If we ever support "`.aar` export" (#6108), *we* will no longer be responsible for creating the `.apk`/`.aab` in all circumstances.  Should we use a more resilient system?

Reply:
> Why parens around `(bundled_assemblies)`? (Yes, it _previously_ had them, but it shouldn't need them _now_…)
> 
Omission

> I also wonder if this is a potential future "bug," _requiring_ that the `.dll` come _before_ the `.pdb`. If we ever support "`.aar` export" (#6108), _we_ will no longer be responsible for creating the `.apk`/`.aab` in all circumstances. Should we use a more resilient system?

Yeah, it's something I've always disliked about this code. I'm going to deal with it in another PR as the fix isn't a one-liner.